raiden_libs/messages/balance_proof.py

```diff
@@ -1,85 +1,112 @@
 import jsonschema
 import time
-import struct
+
+from web3 import Web3
+from eth_utils import is_address, decode_hex
 
 from raiden_libs.messages.message import Message
 from raiden_libs.properties import address_property
 from raiden_libs.messages.json_schema import BALANCE_PROOF_SCHEMA
 from raiden_libs.utils import eth_verify
 
-from eth_utils import is_address, decode_hex
-
 
 class BalanceProof(Message):
+    """ A Balance Proof
+
+    This optionally includse the data for calculating the balance_hash.
+    """
     def __init__(
-            self,
-            channel_id: int,
-            contract_address: str,
-            nonce: int = 0,
-            locksroot: str = '0x%064x' % 0,
-            transferred_amount: int = 0,
-            extra_hash: str = '0x%064x' % 0,
-            chain_id: int = 1,
-            signature: str = None
+        self,
+        channel_identifier: int,
+        token_network_address: str,
+
+        balance_hash: str = '',
+        nonce: int = 0,
+        additional_hash: str = '',
+        chain_id: int = 1,
+        signature: str = '',
+
+        transferred_amount: int = None,
+        locked_amount: int = None,
+        locksroot: str = None,
     ) -> None:
         super().__init__()
-        assert channel_id > 0
-        assert is_address(contract_address)
-        self.channel_id = channel_id
-        self.contract_address = contract_address
+        assert channel_identifier > 0
+        assert is_address(token_network_address)
+
         self._type = 'BalanceProof'
-        self.timestamp = time.time()
+
+        self.channel_identifier = channel_identifier
+        self.token_network_address = token_network_address
+
+        self.balance_hash = balance_hash
+        self.additional_hash = additional_hash
         self.nonce = nonce
-        self.locksroot = locksroot
-        self.transferred_amount = transferred_amount
-        self.extra_hash = extra_hash
-        self.signature = signature
         self.chain_id = chain_id
+        self.signature = signature
+
+        if transferred_amount and locked_amount and locksroot:
+            assert self.hash_balance_data(
```

Comment:
Not sure if we need to do that here. Maybe we should add a `validate` method instead.